### ISSUE TITLE
Failures on 0.17.x with MKL

### ISSUE THREAD

**Issue by Eric89GXL** [scipy/scipy#5621]
Time for this release candidate's set of MKL failures. Do other people test using MKL, or should I try to make a habit of doing this check before each release?

`icc --version` is `icc (ICC) 16.0.1 20151021`

`numpy` on `maintenance/1.10.x` branch, version `1.10.0.dev0+075cc98`, `numpy.test('full')` passing; `scipy.test('full')` gives:
```
>>> scipy.test('full')
Running unit tests for scipy
NumPy version 1.10.0.dev0+075cc98
NumPy relaxed strides checking option: False
NumPy is installed in /home/larsoner/.local/lib/python2.7/site-packages/numpy
SciPy version 0.17.0rc1
SciPy is installed in /home/larsoner/.local/lib/python2.7/site-packages/scipy
Python version 2.7.10 (default, Oct 14 2015, 16:09:02) [GCC 5.2.1 20151010]
nose version 1.3.6

...

======================================================================
FAIL: test_lorentz (test_odr.TestODR)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/larsoner/.local/lib/python2.7/site-packages/scipy/odr/tests/test_odr.py", line 293, in test_lorentz
    3.7798193600109009e+00]),
  File "/home/larsoner/.local/lib/python2.7/site-packages/numpy/testing/utils.py", line 892, in assert_array_almost_equal
    precision=decimal)
  File "/home/larsoner/.local/lib/python2.7/site-packages/numpy/testing/utils.py", line 713, in assert_array_compare
    raise AssertionError(msg)
AssertionError: 
Arrays are not almost equal to 6 decimals

(mismatch 100.0%)
 x: array([  1.000000e+03,   1.000000e-01,   3.800000e+00])
 y: array([  1.430678e+03,   1.339051e-01,   3.779819e+00])

======================================================================
FAIL: test_multi (test_odr.TestODR)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/larsoner/.local/lib/python2.7/site-packages/scipy/odr/tests/test_odr.py", line 190, in test_multi
    0.5101147161764654, 0.5173902330489161]),
  File "/home/larsoner/.local/lib/python2.7/site-packages/numpy/testing/utils.py", line 892, in assert_array_almost_equal
    precision=decimal)
  File "/home/larsoner/.local/lib/python2.7/site-packages/numpy/testing/utils.py", line 713, in assert_array_compare
    raise AssertionError(msg)
AssertionError: 
Arrays are not almost equal to 6 decimals

(mismatch 100.0%)
 x: array([ 4. ,  2. ,  7. ,  0.4,  0.5])
 y: array([ 4.379988,  2.433306,  8.002885,  0.510115,  0.51739 ])

======================================================================
FAIL: test_pearson (test_odr.TestODR)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/larsoner/.local/lib/python2.7/site-packages/scipy/odr/tests/test_odr.py", line 236, in test_pearson
    np.array([5.4767400299231674, -0.4796082367610305]),
  File "/home/larsoner/.local/lib/python2.7/site-packages/numpy/testing/utils.py", line 892, in assert_array_almost_equal
    precision=decimal)
  File "/home/larsoner/.local/lib/python2.7/site-packages/numpy/testing/utils.py", line 713, in assert_array_compare
    raise AssertionError(msg)
AssertionError: 
Arrays are not almost equal to 6 decimals

(mismatch 100.0%)
 x: array([ 1.,  1.])
 y: array([ 5.47674 , -0.479608])

----------------------------------------------------------------------
Ran 21565 tests in 1190.894s

FAILED (KNOWNFAIL=130, SKIP=1811, failures=3)
<nose.result.TextTestResult run=21565 errors=0 failures=3>
>>> 

```
I can also try with latest `numpy` master, but I figured we should try to clear up the `1.10` failures to see if that gets everything working first.

**Comment by ev-br** [comment]
ordqz passes, yay! :-)

> should I try to make a habit of doing this check before each release?

That'd be great. A monthly cron job?

**Comment by ev-br** [comment]
On the actual topic: last non-doc commits to ODR seem to be maintenance cleanups made over a year ago...

**Comment by Eric89GXL** [comment]
Could be it's broken on the latest Intel MKL compiler, then, since that's also been updated by me since the last check. I'll give 0.16.x a try to see if it fails similarly, since I know that did pass last time I ran it (which was with the previous MKL release).

**Comment by ev-br** [comment]
I'd also try to downgrade numpy to 1.6.2 or 1.7.2: it could be due to some
recent tightening in numpy.

It's too easy for me to say "I would" when I'm not doing the work, though.
So I'd rather say thank you for testing it!
24.12.2015 0:52 пользователь "Eric Larson" <notifications@github.com>
написал:

**Comment by Eric89GXL** [comment]
FYI I also have an interpolation failure I wasn't getting with 0.16 that I doubt is related to MKL. I'll try to isolate it to a minimal example.

**Comment by Eric89GXL** [comment]
Same failures on `SciPy version 0.16.2.dev0+728c397`, so it's probably a "Intel Parallel Studio 2016 Update 1" problem. I'll roll back `icc` to 2015 update 6 and see if it works.

**Comment by rgommers** [comment]
The ``ODR`` failures are a known ifort bug that should be solved when you compile with ``-O1``, see gh-3340

**Comment by rgommers** [comment]
Closing this is a duplicate of gh-3340. Thanks @Eric89GXL, good to know that there are no other issues with MKL.

**Comment by Eric89GXL** [comment]
Should we change it to use `-01` by default in the Intel fortran compilers in `numpy` until Intel fixes the problem upstream?

**Comment by rgommers** [comment]
It should be using ``-O1``: https://github.com/numpy/numpy/blob/master/numpy/distutils/fcompiler/intel.py#L172

If you have that line but are still getting ``-O2`` or ``-O3`` then we should have a closer look.

**Comment by Eric89GXL** [comment]
That option is only for the `IntelVisualFCompiler` class, which isn't used on Linux, though. I'm on Ubuntu x84_64, so it uses the `IntelEM64TFCompiler` class IIRC.

**Comment by Eric89GXL** [comment]
For it to work for me I had to add it here:

https://github.com/numpy/numpy/blob/master/numpy/distutils/fcompiler/intel.py#L123

**Comment by rgommers** [comment]
Ah ok. Then I think we fixed it for the reported issue on Windows only. Do you want to send a PR with the change for Linux?

**Comment by Eric89GXL** [comment]
Yeah will do

**Comment by Eric89GXL** [comment]
With `-O1` and `numpy` `1.11.0.dev0+76f2870` the errors above go away, but I do get:
```
ERROR: test_spearmanr (test_mstats_basic.TestCompareWithStats)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/larsoner/.local/lib/python2.7/site-packages/scipy/stats/tests/test_mstats_basic.py", line 974, in test_spearmanr
    r, p = stats.spearmanr(x, y)
  File "/home/larsoner/.local/lib/python2.7/site-packages/scipy/stats/stats.py", line 3314, in spearmanr
    t = rs * np.sqrt((n-2) / ((rs+1.0)*(1.0-rs)))
RuntimeWarning: invalid value encountered in sqrt

======================================================================
ERROR: test_tie1 (test_stats.TestCorrSpearmanrTies)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/larsoner/.local/lib/python2.7/site-packages/scipy/stats/tests/test_stats.py", line 673, in test_tie1
    sr = stats.spearmanr(x, y)
  File "/home/larsoner/.local/lib/python2.7/site-packages/scipy/stats/stats.py", line 3314, in spearmanr
    t = rs * np.sqrt((n-2) / ((rs+1.0)*(1.0-rs)))
RuntimeWarning: invalid value encountered in sqrt
```
Should these be fixed for 0.17? Or did I find another duplicate? :)

**Comment by ev-br** [comment]
This looks like a numpy-dev artifact of turning warnings to errors :-). Would be good to filter out in tests.

**Comment by ev-br** [comment]
https://github.com/scipy/scipy/pull/5609 filters them out *in the library* but I do think it's better to leave the library code alone and filter them out in the tests.

**Comment by ogrisel** [comment]
#5609 has been updated to avoid triggering those warning without adding any new filtering. This PR should be rebased on top of master to benefit from it.

**Comment by ogrisel** [comment]
Actually this is not a PR, but what I meant is that the warnings reported in https://github.com/scipy/scipy/issues/5621#issuecomment-167611301 should not longer be triggered on master.

**Comment by ogrisel** [comment]
If you need the fix for the 0.17 maintenance branch you can cherry-pick 5c187767a0a778b749f999a292f5f680bea85750.
